content/en/docs/tutorials/configuration/pod-sidecar-containers.md

```diff
@@ -0,0 +1,148 @@
+---
+title: Adopting Sidecar Containers
+content_type: tutorial
+weight: 40
+min-kubernetes-server-version: 1.29
+---
+
+<!-- overview -->
+
+This section is relevant for people adopting [sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/) for their workloads.
+
+{{< feature-state feature_gate_name=`SidecarContainers >}}
+
+## {{% heading "objectives" %}}
+
+* Understand the need for sidecar containers
+* Be able to troubleshoot issues with the sidecar containers
+* Understand options to universally "inject" sidecar containers to any workload
+
+
+## {{% heading "prerequisites" %}}
+
+{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}
+
+* This tutorial is relevant for clusters version 1.29+
+
+
+<!-- lessoncontent -->
+
+## Sidecar containers overview
+
+Sidecar containers are the secondary containers that run along with the main
+application container within the same {{< glossary_tooltip text="Pod" term_id="pod" >}}.
+These containers are used to enhance or to extend the functionality of the primary _app
+container_ by providing additional services, or functionality such as logging, monitoring,
+security, or data synchronization, without directly altering the primary application code.
+You can read more at [Sidecar containers concept page](/docs/concepts/workloads/pods/sidecar-containers/).
+
+The concept of sidecar containers is not new and there are multiple implementation of this concept.
+The feature is being used by end users as well as by the sidecar authors who provide ways to "inject"
+sidecar containers to any Pod. Those "injectors" are typically implemented as [mutating webhook](/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook).
+
+While the concept of sidecar containers is not new,
+the native implementation of this feature in Kubernetes, however, is new. And as with every new feature,
+adopting this feature may present certain challenges.
+
+This article explore challenges and solution can be experienced by end users as well as
+by authors of sidecar containers.
+
+## Benefits of a built-in sidecar containers
+
+There are many benefits for the built-in sidecar containers.
+
+1. Built-in sidecar containers can run before Init containers.
+2. It is guaranteed that built-in sidecar containers finish AFTER all the regular containers completion.
```

Comment:
The statement is correct, but it can easily be misinterpreted. At first read, I assumed you meant that the sidecar container would finish successfully with a zero exit code. However, when I tried it by making it run continuously, it finished with a non-zero exit code.

To try this, define a manifest, set the regular container to sleep for say, `10s`, and set the pod’s restart policy to `NEVER`. Lastly, set the restart policy of the init-container to `ALWAYS` and make sure it keeps running by using a while loop that starts it after the sleep of `1s` elapses.
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
spec:
  initContainers:
  - name: sidecar
    image: alpine
    command: ["/bin/sh", "-c", "while true; do echo 'Sidecar is running'; sleep 1; done"] 
    restartPolicy: Always
  containers:
  - name: main
    image: alpine
    command: ["/bin/sh", "-c", "echo 'Main container is running'; sleep 10"]
  restartPolicy: Never
```

You will notice that the main container completes successfully and terminates. However, the sidecar container keeps running until it is forcefully terminated via the `SIGKILL` signal, hence, causing the pod to be at a pod phase of `Succeeded` or a status of `Completed` when you do a `kubectl get pod`.

```suggestion
2. It is guaranteed that built-in sidecar containers are terminated once all the regular containers are completed and terminated.
```

Reply:
Sidecars do get a shutdown signal AIUI, but `["/bin/sh", "-c", "while true; do echo 'Sidecar is running'; sleep 1; done"]` might not have a signal handler that will pass the shutdown signal to `sleep` in a useful way.

ICBW here.

Reply:
I agree. If you use an nginx image for the sidecar and main container, but add a short sleep command only for the main container, both containers shut down successfully with an exit code of 0. 

**pod.yaml**
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
spec:
  initContainers:
  - name: sidecar
    image: nginx:alpine
    restartPolicy: Always
  containers:
  - name: main
    image: alpine
    command: ["/bin/sh", "-c", "echo 'Main container is running'; sleep 10"]
  restartPolicy: Never
```

```yaml
kubectl apply -f pod.yaml

NAME     READY   STATUS    RESTARTS   AGE
my-pod   2/2     Running   0          11s
```

After about 14 seconds
```yaml
kubectl get pod
NAME     READY   STATUS      RESTARTS   AGE
my-pod   0/2     Completed   0          15s
```

```yaml
kubectl logs my-pod sidecar 

2024/07/01 18:06:15 [notice] 1#1: start worker processes
2024/07/01 18:06:15 [notice] 1#1: start worker process 29
2024/07/01 18:06:26 [notice] 1#1: signal 3 (SIGQUIT) received, shutting down
2024/07/01 18:06:26 [notice] 29#29: gracefully shutting down
2024/07/01 18:06:26 [notice] 29#29: exiting
2024/07/01 18:06:26 [notice] 29#29: exit
2024/07/01 18:06:26 [notice] 1#1: signal 17 (SIGCHLD) received from 29
2024/07/01 18:06:26 [notice] 1#1: worker process 29 exited with code 0
2024/07/01 18:06:26 [notice] 1#1: exit
```

```yaml
kubectl describe pod my-pod

main:
    Container ID:  containerd://d67d484d263b88f954b110898b9d0a81b6ff8657143d7a3043f07c4a5a518d3b
    Image:         alpine
....
    Command:
      /bin/sh
      -c
      echo 'Main container is running'; sleep 10
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
```

You can see that they both handle the shutdown signal gracefully.

I am only saying that the statement in the commit saying that it's guaranteed that built-in sidecar containers finish AFTER all the regular containers completion may not be always true.

Reply:
rephrased